### ISSUE TITLE
Freezing and Optimising for Inference

### ISSUE THREAD

**Issue by splitice** [keithito/tacotron#95]
I've been attempting today to create a frozen graph, and then optimise that graph for inference (to evaluate performance for an embedded device).

This model lacks a placeholder node for inference input, is there a reason for this? I'm fairly new to Tensor Flow but most of my models have one, is this a performance optimisation / newbie thing?

I'd be interested to know your thoughs.

**Comment by keithito** [comment]
The input placeholder is here:
https://github.com/keithito/tacotron/blob/master/synthesizer.py#L14

It gets passed to the model rather than being instantiated by the model itself because during training, the input goes through a FIFOQueue, which is more efficient than using a feed_dict.
https://www.tensorflow.org/api_guides/python/threading_and_queues

I think TF 1.4 added some simpler ways of doing this, but this code predates that.

**Comment by splitice** [comment]
Yes thank you I see that now. It's different to how I've done it in the past external to the model.

Thanks for being patient with me.

**Comment by keithito** [comment]
Awesome! 😄

**Comment by MXGray** [comment]
@splitice
@keithito
Would it be possible for you to share your code for freezing and optimizing a trained model with weights?
A guide would also be really helpful.  :)
Thanks in advance!

**Comment by splitice** [comment]
@MXGray It will be a PR when ready. I'm working on a ```export_inference.py``` and a ```eval_inference.py```.

For freezing you just need to add the input nodes to the graph before exporting the graphdef, that's what I was doing wrong.

For optimizing I'm currently being held up behind a tensorflow bug (https://github.com/tensorflow/tensorflow/issues/12157) since I am trying to create a model with reduced memory usage (<400mb).

I'll be looking to quantize the model too.

We are actully just experimenting at this stage though. Hoping that @keithito gets v2 of the paper implemented and it leads to some quality improvements, then this will certainly be the best option for our needs.

**Comment by jeffxtang** [comment]
@keithito thanks for the great repo. when we freeze the graph with your pre-trained checkpoint file, using a script such as https://blog.metaflow.fr/tensorflow-how-to-freeze-a-model-and-serve-it-with-a-python-api-d4f3596b3adc, should we also provide output node name(s)? if so, what should be the name? i saw in synthesizer.py, there's code
```
self.wav_output = audio.inv_spectrogram_tensorflow(self.model.linear_outputs[0])
...
wav = self.session.run(self.wav_output, feed_dict=feed_dict)
```
but can't figure out the output name. Thanks for your help.

**Comment by jeffxtang** [comment]
or @MXGray have you figured out how to freeze the graph? I'm also downloading your pre-trained model on the Nancy dataset and will play with it - thanks!

**Comment by keithito** [comment]
Hi, sorry for the slow reply. The output node in the checkpoint is "model/inference/dense/BiasAdd:0". The way things are currently structured, this is the spectrogram output (i.e. before running Griffin-Lim). As a result, after loading the frozen model, you'll need to attach the Griffin-Lim operations to it -- which I realize is inconvenient. On the other hand, hopefully that part will be obsolete if we can get WaveNet working as a replacement for Griffin-Lim.

**Comment by jeffxtang** [comment]
Thanks @keithito for your reply! I managed to find out the name after debugging tacotron.py (but then was stuck with some errors when loading the graph in iOS so sorry I forgot to update here) and it's good to get your confirmation. 

Your demo script with the web test client works very nice. One thing I noticed is running on a GTX 1070 GPU, the TTS takes only about 1 second even for a pretty long sentence, but on a CPU it takes about 20 seconds. Why does the model inference take so long on a CPU? Guess it won't be feasible running on mobile devices? Any thoughts? Thanks..

**Comment by keithito** [comment]
Griffin-Lim is the slowest part when running on CPU. It's actually faster to use the librosa implementation of Griffin-Lim than the Tensorflow one on the CPU, and you can reduce the number of iterations without much quality loss. I think it might be feasible to get this running on mobile devices -- at least the newer, higher-end ones have some pretty impressive GPUs.

**Comment by jeffxtang** [comment]
Thanks for the tips. I'll definitely give it a try.

**Comment by dreamibor** [comment]
@keithito @jeffxtang @splitice Hi, did you figure out how to freeze the model? I froze the model using the command " python3 /work/tensorflow/tensorflow/python/tools/freeze_graph.py --input_checkpoint="model.ckpt-800000" --clear_devices=true --input_meta_graph=./model.ckpt-800000.meta --output_node_names="model/inference/dense/BiasAdd" --output_graph='graph.pb' --input_binary=true". However, it will runs forever without logging any errors. The input and output nodes I am using are :  
inputs_name = "datafeeder/input_queue_Dequeue:0"
input_lengths_name = "datafeeder/input_queue_Dequeue:1"
output_layer_name = "model/inference/dense/BiasAdd:0"
I then feed them into sess.run(), but it seems the model cannot run through the model. How do you guys freeze the model and use it? Thank you!

**Comment by dreamibor** [comment]
@splitice Did you figure out how to quantize the model?

**Comment by amanmeetgarg** [comment]
@keithito : thanks for sharing the repo
@splitice 

"'For freezing you just need to add the input nodes to the graph before exporting the graphdef, that's what I was doing wrong."

It may come out as naive. Though, how do you add input nodes ? since the inputs otherwise are already input to the FiFoQueueu.  I am confused. Appreciate your help.

Thanks.

**Comment by DanRuta** [comment]
@amanmeetgarg Did you happen to figure this out?

**Comment by amanmeetgarg** [comment]
@DanRuta no solution so far .. for my problem at hand, i ended up using the trained model and imported weights elsewhere.

All the best.

**Comment by splitice** [comment]
Unfortunately the laptop I was doing this work on was stolen in January. Freezing is definitely possible though, I had some data being output but I didn't have a properly trained model (beyond a few thousand steps) to know if I had derped anything up.

**Comment by japita-se** [comment]
Anyone succeed in importing an optimized .pb (I used graph_transform) in synthetizer.py ?

**Comment by japita-se** [comment]
> Did you figure out how to quantize the model?

``
> ./bazel-bin/tensorflow/tools/graph_transforms/transform_graph \
> --in_graph="../TTSDataset_keith/tacotron-60k/freezed.pb" \
> --inputs='import/datafeeder/input_queue' \
> --outputs='model/inference/dense/BiasAdd:0' \
> --out_graph='optimized_ta_keith_60k.pb' \
> --transforms='strip_unused_nodes(type=float, shape="1,224,224,3")fold_constants(ignore_errors=true) fold_batch_norms fold_old_batch_norms quantize_weights  
>  quantize_nodes sort_by_execution_order' \
``

**Comment by japita-se** [comment]
How to load the frozen graph?

def load_graph(self,frozen_graph_filename):
     print(frozen_graph_filename)
     with tf.gfile.GFile(frozen_graph_filename, "rb") as f:
        graph_def = tf.GraphDef()
        graph_def.ParseFromString(f.read())
     with tf.Graph().as_default() as graph:
     # The name var will prefix every node name in graph if needed
        tf.import_graph_def(graph_def, name="")
     return graph

  def load(self, model_path, model_name='tacotron'):
    print('Constructing model: %s' % model_name)
    inputs = tf.placeholder(tf.int32, [1, None], 'inputs')
    input_lengths = tf.placeholder(tf.int32, [1], 'input_lengths')
    with tf.variable_scope('model') as scope:
        self.model = create_model(model_name, hparams)
        self.model.initialize(inputs, input_lengths)
    #  self.wav_output = audio.inv_spectrogram_tensorflow(self.model.linear_outputs[0])

    #print('Loading checkpoint: %s' % checkpoint_path)
    #self.session = tf.Session()
    #self.session.run(tf.global_variables_initializer())
    #saver = tf.train.Saver()
    #saver.restore(self.session, checkpoint_path)
    print('Loading frozen graph for inference: %s' % model_path)
    graph = self.load_graph(model_path)
    self.input = graph.get_tensor_by_name('inputs:0')
    self.input_lengths = graph.get_tensor_by_name('input_lengths:0')
    #self.wav_output = graph.get_tensor_by_name('model/griffinlim/Squeeze:0')
    #OR IF YOU ARE ATTACHING GRIFFINLIM AT INFERENCE TIME USE BELOW
    self.wav_output = audio.inv_spectrogram_tensorflow(graph.get_tensor_by_name('model/inference/dense/BiasAdd:0')[0])
    self.session = tf.Session(graph = graph)
